### ISSUE TITLE
SystemError: unknown opcode when trying to execute a package in python 3.9 that was obfuscated in python 3.8

### ISSUE THREAD

**Issue by jdominguezg89** [dashingsoft/pyarmor#691]
When I try to import and run the package in Jupyter Notebook running in a python 3.9 environment I get the following error:
> XXX lineno: 1, opcode: 53
XXX lineno: 16, opcode: 88
XXX lineno: 3, opcode: 88
XXX lineno: 12, opcode: 88

```
SystemError                               Traceback (most recent call last)
/opt/anaconda3/lib/python3.9/site-packages/testpackage/src/__init__.py in <module>

/opt/anaconda3/lib/python3.9/site-packages/testpackage/src/resources/__init__.py in <module>

/opt/anaconda3/lib/python3.9/site-packages/testpackage/src/resources/__init__.py in <module>

SystemError: unknown opcode
```

I was doing some research and found that in the [pyarmor change logs](https://pyarmor.readthedocs.io/en/latest/change-logs.html#id33) this error had already been corrected.

I also checked the [reported bug](https://github.com/dashingsoft/pyarmor/issues/357) to see if I could find a guide but nothing either.

I'm obfuscating the package by executing the command bellow:

```commandline
pyarmor obfuscate --platform darwin.aarch64,darwin.x86_64,ios.aarch64,linux.aarch32,linux.aarch64,linux.armv6,linux.armv7,linux.x86,linux.x86_64,musl.x86_64,windows.x86,windows.x86_64 --recursive --output ./obfuscated ./"$FOLDER"/testpackage/__init__.py;
```

After that I'm building a pyarmored wheel and uploading the package to the test pypi.org environment.

I'm using the latest version of pyarmor.

Any guide or suggestion on how I can solve this problem?

Very grateful in advance.

**Comment by jondy** [comment]
Obfuscate the scripts for one platform, test it work or not.

**Comment by jdominguezg89** [comment]
I already tried to obfuscate the code for a single platform in `python 3.8.12`, with this command:

`pyarmor obfuscate --platform darwin.x86_64 --recursive --output ./obfuscated ./"$FOLDER"/testpackage/__init__.py;`

But then I try to import the package and run it in a python 3.9 environment as mentioned above, and the same error persists, what else can I do to solve this?

**Comment by jdominguezg89** [comment]
Please some workaround to fix this but using several platforms? i really need to solve it!

**Comment by jondy** [comment]
The solution has been documentation, please read documentation to find the solution.

**Comment by jdominguezg89** [comment]
Hi @jondy, thank you very much for your answer, but the thing is that I still can't solve the problem.

To begin with, I have read the pyarmor documentation again, with which I have understood much better how it works, emphasizing the chapters [Understanding Obfuscated Scripts](https://pyarmor.readthedocs.io/en/latest/understand-obfuscated-scripts.html), [When Things Go Wrong](https://pyarmor.readthedocs.io/en/latest/questions.html), [Support Platforms](https://pyarmor.readthedocs.io/en/latest/platforms.html) and [Advanced Topics](https://pyarmor.readthedocs.io/en/latest/advanced.html), because I think they are the most important to read and keep in mind for decent obfuscation.

I also checked the Change Logs and saw that in the last update ([7.3.6](https://pyarmor.readthedocs.io/en/latest/change-logs.html#id4)) some change was made related to this topic.

The truth is that I have not found the solution that you say in the previous comment that is documented, I am sorry if I am a fool for not finding it.

Try executing the command with the `-d` option and saving the logs to a file `>../pyarmor_log.txt 2>&1` to see if anything goes wrong, but the execution succeeded.

Execution command:
```commandline
pyarmor -d obfuscate --platform darwin.aarch64,darwin.x86_64 --recursive -O ./obfuscated ./"$FOLDER"/testpackage/__init__.py >../pyarmor_log.txt 2>&1;
```
 Again, after run the import command in Jupyter Notebook running on python 3.9 the same error appears:
> XXX lineno: 1, opcode: 53
XXX lineno: 16, opcode: 88
XXX lineno: 3, opcode: 88
XXX lineno: 12, opcode: 88

```
SystemError                               Traceback (most recent call last)
/opt/anaconda3/lib/python3.9/site-packages/testpackage/src/__init__.py in <module>

/opt/anaconda3/lib/python3.9/site-packages/testpackage/src/resources/__init__.py in <module>

/opt/anaconda3/lib/python3.9/site-packages/testpackage/src/resources/__init__.py in <module>

SystemError: unknown opcode
```

But if i do that in Jupyter running over python 3.8 everything works as expected.

## Obfuscating using `--advanced 2`

Also try obfuscating the package using the `--advanced 2` option to enable super mode:

```commandline
pyarmor -d obfuscate --advanced 2 --platform darwin.aarch64,darwin.x86_64 --recursive -O ./obfuscated ./"$FOLDER"/testpackage/__init__.py >../pyarmor_log.txt 2>&1;
```

In this case, i've got a different error:

```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
/var/folders/sr/yt11h05x0qzgh930_fmvz6040000gn/T/ipykernel_14257/840950783.py in <module>
----> 1 from testpackage import tp
      2 import pandas as pd

/opt/anaconda3/lib/python3.9/site-packages/testpackage/__init__.py in <module>
----> 1 from pytransform import pyarmor
      2 pyarmor(__name__, __file__, b'\x50\x59\x41\x52\x4d\x4f\x52\x00\x00\x03\x08\x00\x55\x0d\x0d\x0a\x09\x30\xe0\x02\x01\x00\x00\x00\x01\x00\x00\x00\x40\x00\x00\x00\x85\x02\x00\x00\x0b\x00\x00\x18\x7e\xab\xf6\xce\x45\x11\x46\x18\x00\x01\x87\xfb\xb0\xbe\xb6\x39\x00\x00\x00\x00\x00\x00\x00\x00\x1d\x62\x21\xa1\x8a\x43\x08\x63\x89\x0e\x92\xec\xc9\xc4\x8c\xff\xdf\x98\x64\x79\x96\xe7\x73\xa0\xa0\x82\x85\xc1\x41\xbe\xde\xe6\x69\x92\x17\xae\x46\x84\x0b\x12\xf0\xd9\xea\x8e\xfd\xda\x98\x0f\x21\x99\xa7\x2c\x21\xe2\x86\x98\x64\x3c\x85\x87\x65\x65\x14\x12\x1d\x07\x6e\x93\xf3\x44\xbc\xd7\xe4\xfa\x1d\xe6\x99\x59\x46\xce\x95\xdb\x3d\x4f\xad\x5c\x94\x41\x38\xae\x3b\xe8\x98\x00\x55\x5f\x4f\x45\x98\xee\x02\x15\xf9\xdd\xd9\x72\xef\x8a\xb5\x26\x4c\x32\x0d\x41\x45\x8d\x7d\x3d\xe5\xa3\x57\xba\x52\x53\x3c\x12\x4e\x5f\x1b\x31\x45\x91\x70\x8a\xb2\x88\x8d\x64\x82\x88\xa2\xca\x73\x61\x0e\xbb\x2a\xcb\x76\x19\x82\x35\x45\xf7\x4b\x6d\x45\xb5\x59\xb6\xd7\x72\x79\x65\xcc\x79\xe0\x63\xb0\xf8\x92\xd7\x7b\xa2\xa8\x89\x5b\x29\x96\x21\x6f\xa6\x04\x51\x4a\x80\x44\x37\x37\xce\x26\xbd\xff\xe7\x44\x97\xab\x8b\x82\x99\x8d\x12\x2b\xda\x7a\x3a\x9b\x59\x21\x0e\x32\xeb\x85\xdb\x6f\xd9\x02\x50\x6b\x7a\xd7\xe7\x09\xa0\xea\x7d\x18\x01\xba\xdc\x99\xe7\x19\xea\xc5\xc0\x2f\xb7\x21\x2f\x18\x0e\x23\x6a\x44\xa4\x3e\x5c\x08\x82\x6e\x99\x95\xf1\xcb\x90\xc7\xbd\xb9\x94\x0c\xac\x7a\x77\xe1\xfd\x2d\xab\x2c\x66\x15\x83\x36\xcf\x75\x75\x4d\x11\x97\x99\x3c\x1f\x51\x4b\x84\x77\x36\xbc\x3e\x05\x62\x9f\x03\xcc\x1d\x13\xb7\x1a\x1f\xbf\x88\x9a\x02\xc0\x8a\x64\xe3\x75\x26\xf8\xae\x1f\xec\x65\xbf\xd7\x29\x28\x41\xde\x02\x8b\x13\x9f\x0b\xdb\xcd\x4a\x85\x68\xca\x52\x01\xe8\x7f\xf4\x1f\xcf\xc0\x70\xd7\xb8\xdb\x51\x98\x17\x40\x5a\x2c\x33\x6a\x5b\x85\x56\x3a\xba\xda\x3f\xe9\x49\x94\x32\xbf\xfe\xde\xf2\x57\x06\xb4\x4e\x3a\x64\x1d\xe0\xc4\xd0\x54\x41\x39\x82\x42\xe1\x2f\xa0\x98\xda\xbf\x43\x39\xa1\xa1\x4d\x9d\x7d\xdb\xa5\xa1\x5a\x5d\x1d\xc4\xce\x7f\x4b\x65\x08\xf6\x7c\xe6\x69\x59\xc5\x05\xa0\x5e\x34\xaf\xd9\x8c\x50\x42\x2b\x4c\x87\x07\xb0\x79\x19\xd6\x93\x76\x2a\xa0\xcc\x28\x14\x90\x84\xe9\x7b\x8e\x72\xb1\xb2\x07\x9f\xa7\x4c\xcf\x1f\x1f\x19\x16\x03\x7a\x06\x76\x6e\x2c\xe4\xa5\xf8\x8e\x8d\x7d\x84\x0d\xc3\x04\x54\x76\x15\x80\x9d\xbd\x55\x87\xe4\xc6\xd8\x39\x5f\xaa\x84\x33\xc6\xbb\x0b\x0e\xf1\xdb\x2f\xd1\x1b\xff\x6b\x2b\xc5\x24\xa7\x5d\x28\xe9\x79\x29\x62\x00\x5a\x4a\x7b\x52\xe4\x81\xd7\xa6\x17\xa3\xaf\xe2\x5c\xf8\x05\x96\x46\xd1\x5d\xf1\x83\xb6\xcf\x5b\x6c\x98\xa5\xf8\xf4\xb6\x72\xe5\x33\xed\xc9\x8c\x77\xf7\x17\x5b\xb0\x53\x76\xb5\x55\xa6\xb2\xd4\xf1\x24\xec\x01\x9d\x89\x2f\xa9\xb7\xd1\x2f\xcc\x42\xe9\x2e\x22\x55\xd5\x4d\x05\xde\x97\x74\x43\x9c\xb5\x11\x2f\x75\x9f\x4c\x69\xa5\xd2\xa3\xee\x62\x6a\x3f\xaa\x06\x22\xe7\x26\x85\x47\xcc\x31\x2b\xaa\x6d\x30\x13\xd9\x6f\x6c\x13\xa0\xed\x6e\x9f\x99\x19\xc7\xec\x4b\x99\x80\x50\xf3\xf6\x7a\xe4\xf0\x6b\x70\x47\xa7\x9f\x99\x38\x13\xaf\xf2\xd1\xcd\xeb\xa7\x20\xc3', 2)

ModuleNotFoundError: No module named 'pytransform'
```

## In summary

I need to obfuscate the code in python 3.8, so that it can be run on various platforms and on python versions greater than or equal to python 3.8, after end users install it from pypi.org.

Any prompt help will be appreciated.

**Comment by jdominguezg89** [comment]
> The solution has been documentation, please read documentation to find the solution.

Hi @jondy, thank you very much for your answer, but the thing is that I still can't solve the problem.

To begin with, I have read the pyarmor documentation again, with which I have understood much better how it works, emphasizing the chapters [Understanding Obfuscated Scripts](https://pyarmor.readthedocs.io/en/latest/understand-obfuscated-scripts.html), [When Things Go Wrong](https://pyarmor.readthedocs.io/en/latest/questions.html), [Support Platforms](https://pyarmor.readthedocs.io/en/latest/platforms.html) and [Advanced Topics](https://pyarmor.readthedocs.io/en/latest/advanced.html), because I think they are the most important to read and keep in mind for decent obfuscation.

I also checked the Change Logs and saw that in the last update ([7.3.6](https://pyarmor.readthedocs.io/en/latest/change-logs.html#id4)) some change was made related to this topic.

The truth is that I have not found the solution that you say in the previous comment that is documented, I am sorry if I am a fool for not finding it.

Try executing the command with the `-d` option and saving the logs to a file `>../pyarmor_log.txt 2>&1` to see if anything goes wrong, but the execution succeeded.

Execution command:
```commandline
pyarmor -d obfuscate --platform darwin.aarch64,darwin.x86_64 --recursive -O ./obfuscated ./"$FOLDER"/testpackage/__init__.py >../pyarmor_log.txt 2>&1;
```
 Again, after run the import command in Jupyter Notebook running on python 3.9 the same error appears:
> XXX lineno: 1, opcode: 53
XXX lineno: 16, opcode: 88
XXX lineno: 3, opcode: 88
XXX lineno: 12, opcode: 88

```
SystemError                               Traceback (most recent call last)
/opt/anaconda3/lib/python3.9/site-packages/testpackage/src/__init__.py in <module>

/opt/anaconda3/lib/python3.9/site-packages/testpackage/src/resources/__init__.py in <module>

/opt/anaconda3/lib/python3.9/site-packages/testpackage/src/resources/__init__.py in <module>

SystemError: unknown opcode
```

But if i do that in Jupyter running over python 3.8 everything works as expected.

## Obfuscating using `--advanced 2`

Also try obfuscating the package using the `--advanced 2` option to enable super mode:

```commandline
pyarmor -d obfuscate --advanced 2 --platform darwin.aarch64,darwin.x86_64 --recursive -O ./obfuscated ./"$FOLDER"/testpackage/__init__.py >../pyarmor_log.txt 2>&1;
```

In this case, i've got a different error:

```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
/var/folders/sr/yt11h05x0qzgh930_fmvz6040000gn/T/ipykernel_14257/840950783.py in <module>
----> 1 from testpackage import tp
      2 import pandas as pd

/opt/anaconda3/lib/python3.9/site-packages/testpackage/__init__.py in <module>
----> 1 from pytransform import pyarmor
      2 pyarmor(__name__, __file__, b'\x50\x59\x41\x52\x4d\x4f\x52\x00\x00\x03\x08\x00\x55\x0d\x0d\x0a\x09\x30\xe0\x02\x01\x00\x00\x00\x01\x00\x00\x00\x40\x00\x00\x00\x85\x02\x00\x00\x0b\x00\x00\x18\x7e\xab\xf6\xce\x45\x11\x46\x18\x00\x01\x87\xfb\xb0\xbe\xb6\x39\x00\x00\x00\x00\x00\x00\x00\x00\x1d\x62\x21\xa1\x8a\x43\x08\x63\x89\x0e\x92\xec\xc9\xc4\x8c\xff\xdf\x98\x64\x79\x96\xe7\x73\xa0\xa0\x82\x85\xc1\x41\xbe\xde\xe6\x69\x92\x17\xae\x46\x84\x0b\x12\xf0\xd9\xea\x8e\xfd\xda\x98\x0f\x21\x99\xa7\x2c\x21\xe2\x86\x98\x64\x3c\x85\x87\x65\x65\x14\x12\x1d\x07\x6e\x93\xf3\x44\xbc\xd7\xe4\xfa\x1d\xe6\x99\x59\x46\xce\x95\xdb\x3d\x4f\xad\x5c\x94\x41\x38\xae\x3b\xe8\x98\x00\x55\x5f\x4f\x45\x98\xee\x02\x15\xf9\xdd\xd9\x72\xef\x8a\xb5\x26\x4c\x32\x0d\x41\x45\x8d\x7d\x3d\xe5\xa3\x57\xba\x52\x53\x3c\x12\x4e\x5f\x1b\x31\x45\x91\x70\x8a\xb2\x88\x8d\x64\x82\x88\xa2\xca\x73\x61\x0e\xbb\x2a\xcb\x76\x19\x82\x35\x45\xf7\x4b\x6d\x45\xb5\x59\xb6\xd7\x72\x79\x65\xcc\x79\xe0\x63\xb0\xf8\x92\xd7\x7b\xa2\xa8\x89\x5b\x29\x96\x21\x6f\xa6\x04\x51\x4a\x80\x44\x37\x37\xce\x26\xbd\xff\xe7\x44\x97\xab\x8b\x82\x99\x8d\x12\x2b\xda\x7a\x3a\x9b\x59\x21\x0e\x32\xeb\x85\xdb\x6f\xd9\x02\x50\x6b\x7a\xd7\xe7\x09\xa0\xea\x7d\x18\x01\xba\xdc\x99\xe7\x19\xea\xc5\xc0\x2f\xb7\x21\x2f\x18\x0e\x23\x6a\x44\xa4\x3e\x5c\x08\x82\x6e\x99\x95\xf1\xcb\x90\xc7\xbd\xb9\x94\x0c\xac\x7a\x77\xe1\xfd\x2d\xab\x2c\x66\x15\x83\x36\xcf\x75\x75\x4d\x11\x97\x99\x3c\x1f\x51\x4b\x84\x77\x36\xbc\x3e\x05\x62\x9f\x03\xcc\x1d\x13\xb7\x1a\x1f\xbf\x88\x9a\x02\xc0\x8a\x64\xe3\x75\x26\xf8\xae\x1f\xec\x65\xbf\xd7\x29\x28\x41\xde\x02\x8b\x13\x9f\x0b\xdb\xcd\x4a\x85\x68\xca\x52\x01\xe8\x7f\xf4\x1f\xcf\xc0\x70\xd7\xb8\xdb\x51\x98\x17\x40\x5a\x2c\x33\x6a\x5b\x85\x56\x3a\xba\xda\x3f\xe9\x49\x94\x32\xbf\xfe\xde\xf2\x57\x06\xb4\x4e\x3a\x64\x1d\xe0\xc4\xd0\x54\x41\x39\x82\x42\xe1\x2f\xa0\x98\xda\xbf\x43\x39\xa1\xa1\x4d\x9d\x7d\xdb\xa5\xa1\x5a\x5d\x1d\xc4\xce\x7f\x4b\x65\x08\xf6\x7c\xe6\x69\x59\xc5\x05\xa0\x5e\x34\xaf\xd9\x8c\x50\x42\x2b\x4c\x87\x07\xb0\x79\x19\xd6\x93\x76\x2a\xa0\xcc\x28\x14\x90\x84\xe9\x7b\x8e\x72\xb1\xb2\x07\x9f\xa7\x4c\xcf\x1f\x1f\x19\x16\x03\x7a\x06\x76\x6e\x2c\xe4\xa5\xf8\x8e\x8d\x7d\x84\x0d\xc3\x04\x54\x76\x15\x80\x9d\xbd\x55\x87\xe4\xc6\xd8\x39\x5f\xaa\x84\x33\xc6\xbb\x0b\x0e\xf1\xdb\x2f\xd1\x1b\xff\x6b\x2b\xc5\x24\xa7\x5d\x28\xe9\x79\x29\x62\x00\x5a\x4a\x7b\x52\xe4\x81\xd7\xa6\x17\xa3\xaf\xe2\x5c\xf8\x05\x96\x46\xd1\x5d\xf1\x83\xb6\xcf\x5b\x6c\x98\xa5\xf8\xf4\xb6\x72\xe5\x33\xed\xc9\x8c\x77\xf7\x17\x5b\xb0\x53\x76\xb5\x55\xa6\xb2\xd4\xf1\x24\xec\x01\x9d\x89\x2f\xa9\xb7\xd1\x2f\xcc\x42\xe9\x2e\x22\x55\xd5\x4d\x05\xde\x97\x74\x43\x9c\xb5\x11\x2f\x75\x9f\x4c\x69\xa5\xd2\xa3\xee\x62\x6a\x3f\xaa\x06\x22\xe7\x26\x85\x47\xcc\x31\x2b\xaa\x6d\x30\x13\xd9\x6f\x6c\x13\xa0\xed\x6e\x9f\x99\x19\xc7\xec\x4b\x99\x80\x50\xf3\xf6\x7a\xe4\xf0\x6b\x70\x47\xa7\x9f\x99\x38\x13\xaf\xf2\xd1\xcd\xeb\xa7\x20\xc3', 2)

ModuleNotFoundError: No module named 'pytransform'
```

## In summary

I need to obfuscate the code in python 3.8, so that it can be run on various platforms and on python versions greater than or equal to python 3.8, after end users install it from pypi.org.

Any prompt help will be appreciated.

**Comment by jondy** [comment]
If you have read the common solutions in the When Thing Go Wrong page, you should know it's impossible for pyarmor to obfuscate in Python3.9 then run it in python3.8.

**Comment by jdominguezg89** [comment]
> 



> If you have read the common solutions in the When Thing Go Wrong page, you should know it's impossible for pyarmor to obfuscate in Python3.9 then run it in python3.8.

Hello @jondy, friend, in fact I read the page When Thing Go Wrong as I mentioned in the previous message, and yes, I know that it is not possible to obfuscate the code in python 3.9 and then run it in python 3.8, but what I am trying to do is it just the opposite, obfuscate it in python 3.8 and run it in python 3.8, 3.9, 3.10 or higher, is that not possible either?

**Comment by jondy** [comment]
At this time, the answer is No.

In the advnaced usage, there is a way to merge different python version's scripts so that the obfuscated scripts could run in the different python versions.

**Comment by jdominguezg89** [comment]
> At this time, the answer is No. In the advnaced usage, there is a way to merge different python version's scripts so that the obfuscated scripts could run in the different python versions.

Thanks @jondy, i'll give it a try, and let you know.

**Comment by jdominguezg89** [comment]
Hi @jondy, I have been trying to merge the obfuscated files from the different obfuscated folders but I get the following error when running the merge command:

```
INFO     Merge './obfuscated_base/ ./obfuscated_env310/ ./obfuscated_env3613/ ./obfuscated_env3711/ ./obfuscated_env3812/ '...
INFO     Merging obfuscated scripts...
INFO     Find scripts in the path obfuscated_base/ ./obfuscated_env310/ ./obfuscated_env3613/ ./obfuscated_env3711/ ./obfuscated_env3812/ 
INFO     Merging obfuscated scripts OK
INFO     Merging runtime files...
DEBUG    Found runtimes: []
Traceback (most recent call last):
  File "/opt/anaconda3/envs/env3812/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/anaconda3/envs/env3812/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/anaconda3/envs/env3812/lib/python3.8/site-packages/pyarmor/helper/merge.py", line 298, in <module>
    main()
  File "/opt/anaconda3/envs/env3812/lib/python3.8/site-packages/pyarmor/helper/merge.py", line 287, in main
    merge_runtimes(args.path, output)
  File "/opt/anaconda3/envs/env3812/lib/python3.8/site-packages/pyarmor/helper/merge.py", line 170, in merge_runtimes
    raise RuntimeError('No runtime files found')
RuntimeError: No runtime files found
```

What I am doing is the following:

#### 1. Obfuscate the package for every python version

```commandline
pyarmor -d obfuscate --platform darwin.aarch64,darwin.x86_64 --recursive -O ./obfuscated"_$env" ./"$FOLDER"/testpackage/__init__.py;
```
> Note: I've tried also by adding the `--no-cross-protection` option just in case.

#### 2. Merging every obfuscated directory
```commandline
python -m pyarmor.helper.merge -d -O ./merge_dist ./obfuscated_base/ ./obfuscated_env310/ ./obfuscated_env3613/ ./obfuscated_env3711/ ./obfuscated_env3812/
```

But after running the merge process, it doesn't finish successfully and I get the above error. Please your help with this new bug.

Thank you very much in advance.

**Comment by jdominguezg89** [comment]
@jondy, now everything works great!! Thanks a lot
